client/uma_token_source.go

```diff
@@ -0,0 +1,173 @@
+package client
+
+import (
+	b64 "encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/bitrise-io/bitrise-oauth/config"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+const (
+	umaGrantType        = "urn:ietf:params:oauth:grant-type:uma-ticket"
+	umaClaimTokenFormat = "urn:ietf:params:oauth:token-type:jwt"
+
+	grantType        = "grant_type"
+	claimToken       = "claim_token"
+	claimTokenFormat = "claim_token_format"
+	clientID         = "client_id"
+	clientSecret     = "client_secret"
+	permission       = "permission"
+	audience         = "audience"
+
+	contentType    = "Content-Type"
+	formURLEncoded = "application/x-www-form-urlencoded"
+)
+
+type tokenJSON struct {
+	AccessToken  string `json:"access_token"`
+	TokenType    string `json:"token_type"`
+	RefreshToken string `json:"refresh_token"`
+	ExpiresIn    int32  `json:"expires_in"`
+}
+
+func (e *tokenJSON) expiry(baseTime time.Time) time.Time {
+	if v := e.ExpiresIn; v != 0 {
+		return baseTime.Add(time.Duration(v) * time.Second)
+	}
+
+	return time.Now()
+}
+
+// UMATokenSource represents an UMA token source.
+type UMATokenSource interface {
+	Token(claim interface{}, permisson []Permission, audienceConfig config.AudienceConfig) (*oauth2.Token, error)
+}
+
+type umaTokenSource struct {
+	config clientcredentials.Config
+}
+
+// NewUMATokenSource returns a new UMA token source.
+func newUMATokenSource(config clientcredentials.Config) umaTokenSource {
+	return umaTokenSource{
+		config: config,
+	}
+}
+
+// Token returns a new UMA token upon each invocation.
+func (tokenSource umaTokenSource) Token(claim interface{}, permisson []Permission, audienceConfig config.AudienceConfig) (*oauth2.Token, error) {
+	encodedClaim, err := encodeClaim(claim)
+	if err != nil {
+		return nil, err
+	}
+
+	request, err := tokenSource.newTokenRequest(encodedClaim, permisson, audienceConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := sendRequest(request)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := extractResponseBody(response)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := extractTokenFromBody(body, time.Now())
+	if err != nil {
+		return nil, err
+	}
+
+	return token, nil
+}
+
+func encodeClaim(claim interface{}) (string, error) {
+	bytes, err := json.Marshal(claim)
+	if err != nil {
+		return "", err
+	}
+
+	return b64.StdEncoding.EncodeToString(bytes), nil
+}
+
+func (tokenSource umaTokenSource) newTokenRequest(encodedClaim string, permisson []Permission, audienceConfig config.AudienceConfig) (*http.Request, error) {
+	v := url.Values{}
+
+	v.Set(grantType, umaGrantType)
+	v.Set(claimToken, encodedClaim)
+	v.Set(claimTokenFormat, umaClaimTokenFormat)
+	v.Set(clientID, tokenSource.config.ClientID)
+	v.Set(clientSecret, tokenSource.config.ClientSecret)
+
+	for _, p := range permisson {
+		v.Set(permission, p.requestParam())
```

Comment:
We should use `v.Add` here, not to override the previous permission. Please change line 119 as well.

Reply:
You are right, nice catch!